active/003-seq-feature.md

```diff
@@ -0,0 +1,126 @@
+>>> f.wref   # get weak ref to the interval metadata
+>>> f.function   # get the feature info
+```
+Note, in the above example, the interval `1` is shorthand for `(1, 2)`.
+
+`update(**kwargs)`
+- `**kwargs`: List of attributes that will be modified for the given `BoundFeature` object.
+- Updates the corresponding entries in the `IntervalMetadata`, if intervals are modified, through the weakref
+
+```python
+f = BoundFeature(name='sagA', function='transport')
+f.update(function='toxin')
+f.update(intervals=[(1, 2)])
+```
+
+
+## `IntervalMetadata` object
+### Attributes
+* `features`: a list of `BoundFeature` objects
+* `_staled`: boolean. Whenever any intervals of any member feature is modified, this is set to True, indicating `_intervaltree` needs to be updated.
+* `_intervaltree` (implemented as property): `IntervalTree` object created from all the intervales in the `features`. When this attributes is accessed by any methods (like `query` below), it checks `_staled` to decide whether to re-create from all intervals before it returns the interval tree. This lazy update saves computation.
+
+   This is implemented as bx-python `IntervalTree` object.  The keys correspond to intervals and the values correspond to a single `BoundFeature` object. 
+
+   Some benchmarks have been made for various IntervalTree data structures - in this [benchmark](https://gist.github.com/shoyer/c939325f509d7c027949), bx-python was determined to have one of the fastest query times.
+
+### Methods
+`__init__(self, features=None)`
+- `features` : list of `BoundFeature` objects.
+- Called to initialize the `IntervalMetadata` object
+- The weakrefs will be updated here to link each `BoundFeature` object to the current `IntervalMetadata` object
+
+`reverse_complement(length)`
+- `length` : int.  The length of the whole sequence.  This is required when swapping coordinates
+- This performs a reverse complement on all the coordinates with in IntervalTree.
+- Set `_staled` to True
+
+```python
+   >>> interval_metadata = IntervalMetadata()
+   >>> interval_metadata.add(Feature(gene='sagB', cds=True, intervals=[(3, 5)])
+   >>> iv = interval_metadata.reverse_complement(length=10)
+   >>> f = iv.query(gene='sagB')
+   >>> f.intervals
+   [(5, 7)]
+```
+
+`add(feature)`
+- `feature` : `BoundFeature`. Add a new feature into the `IntervalMetadata` object
+- This allows for a single feature (include those that have non-continugous features) to be added into the `IntervalMetadata` object.
+- The weakref of the added `BoundFeature` will be updated
+- set `_staled` to True.
+```python
+   interval_metadata = IntervalMetadata()
+   interval_metadata.add(BoundFeature([1, (4, 7)], gene='sagA', function='toxin'))
+   interval_metadata.add(BoundFeature([(3, 5)], gene='sagB', function='toxin'), )
+```
+
+`drop(feature)`
+- Deletes a `BoundFeature`.
+- Set `_staled` to True
+
+`query(*args, **kwargs)`
+- `*args` : an iterable of interval tuples to search for features
```

Comment:
Instead of a variable number of args, it could be useful to accept an iterable of tuples so that `BoundFeature.intervals` can be plugged directly into this without unpacking.

How does searching work for enforced attributes such as `boundaries`? Perhaps it's clearer to make a distinction in the signature between arbitrary and enforced attributes if the search rules differ.